src/NUnitFramework/framework/Attributes/FixtureLifeCycleAttribute.cs

```diff
@@ -0,0 +1,75 @@
+// ***********************************************************************
+// Copyright (c) 2019 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using NUnit.Framework.Interfaces;
+using NUnit.Framework.Internal;
+
+namespace NUnit.Framework
+{
+    /// <summary>
+    /// Specify the lifecycle of a Fixture
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Assembly | AttributeTargets.Class, AllowMultiple = false, Inherited = true)]
+    public sealed class FixtureLifeCycleAttribute : PropertyAttribute
```

Comment:
It seems to me that making this a `PropertyAttribute` rather than setting the context complicates things more than is necessary. Commands, which are intended to be as independent of the tests as reasonably possible, have to examine the test, sometimes even the parent tests, in order to do different things based on life cycle. In the original design, which has admittedly changed on an adhoc basis as problems arose, all the "intelligence" about tests was incorporated in the point where the command chains were generated, not internally in the commands themselves. Each command was intended to do one thing. What we have here is a new "thing" to do, so maybe we need a new command.

Reply:
To clarify... I'm OK with a PropertyAtribute if the property is interrogated when the command chains are created, to control what commands are in the chain. Less OK with adding another branch inside an existing command. I'd like to hear what others think about this before deciding, however.

Reply:
Alternative to a `PropertyAttribute` is one that implements `IApplyToContext`. Then the commands would still have branches, but they would be context-based and would not require interrogating the test that created the command.

Reply:
this makes sense, i switched to a simple NUnitAttribute + IApplyToContext, but im having trouble with two parts

1)
right now i'm using:
```
            var fixture = TestBuilder.MakeFixture(typeof(LifeCycleFixtureTestCountIsAlwaysOne));
            var attr = new FixtureLifeCycleAttribute(LifeCycle.SingleInstance);
```
and attr.ApplyToTest(fixture); to apply the attribute to the fixture, but now im not sure how i can apply a context attribute to a fixture.. i'm probably missing something since i haven't seen it done anywhere in the test code

2)
 i still have to change the command chain construction, in SimpleWorkItem MakeTestCommand() since the IApplyToContext is already executed and i can identify the LifeCycle attribute.
but i still need to make the same change (i assume) in  MakeOneTimeSetUpCommand,  but the context is not applied there yet.. so i have no way to know if the fixture is marked as an InstancePerTestCase fixture

Reply:
As you know, `IApplyToContext` has one method, `ApplyToContext`. That method modifies the context of the test where it's applied. NUnit's standard command chain call `ApplyToContext` at the appropriate place, when you have used an attribute that implements the interface.

I assume that your attribute will set a new property in the `TestExecutionContext`. Once that is set, all the contexts of any descendant test cases will have the property set. That leaves nothing for you to do except to recognize the context in your own commands.

Many of NUnit's own attributes work this way. See, for example, `SetCultureAttribute` and `SingleThreadedAttribute`. Our tests for such attributes are all in `AppllyToContextTests`.

Reply:
right now there's code in in `CompositeWorkItem` `MakeOneTimeSetUpCommand()`  that adds `ConstructFixtureCommand` to the chain:

```csharp
private TestCommand MakeOneTimeSetUpCommand(List<SetUpTearDownItem> setUpTearDown, List<TestActionItem> actions)
{
            TestCommand command = new EmptyTestCommand(Test);

            // Add Action Commands
            int index = actions.Count;
            while (--index >= 0)
                command = new BeforeTestActionCommand(command, actions[index]);

            if (Test.TypeInfo != null)
            {
                // Build the OneTimeSetUpCommands
                foreach (SetUpTearDownItem item in setUpTearDown)
                    command = new OneTimeSetUpCommand(command, item);

                // Construct the fixture if necessary
                if (!Test.TypeInfo.IsStaticClass)
                    command = new ConstructFixtureCommand(command); 
            }

            // Prefix with any IApplyToContext items from attributes
            foreach (var attr in Test.GetCustomAttributes<IApplyToContext>(true))
                command = new ApplyChangesToContextCommand(command, attr); 

            return command;
}

```

this is called from `PerformWork()` **InitializeSetUpAndTearDownCommands**
```
        protected override void PerformWork()
        {
            if (!CheckForCancellation())
                if (Test.RunState == RunState.Explicit && !Filter.IsExplicitMatch(Test))
                    SkipFixture(ResultState.Explicit, GetSkipReason(), null);
                else
                    switch (Test.RunState)
                    {
                        default:
                        case RunState.Runnable:
                        case RunState.Explicit:
                            // Assume success, since the result will otherwise
                            // default to inconclusive.
                            Result.SetResult(ResultState.Success);

                            if (Children.Count > 0)
                            {
                                InitializeSetUpAndTearDownCommands();

                                PerformOneTimeSetUp();
```
all the `IApplyToContext`'s attributes `ApplyToContext` are being executed by `PerformOneTimeSetUp` which happens after **InitializeSetUpAndTearDownCommands**
so at this point, Context.LifeCycle (my new property - added to TestExecutionContext) is not applied yet

Reply:
@CharliePoole, you've been following this PR more closely than I have. Do you think that the new instance per fixture command should be created in the `MakeOneTimeSetUpCommand` rather than down in the `SimpleWorkItem`?

Reply:
this.. is a good question, i agree on the second approach, as soon as we come to an agreement ill add the test cases and implementation for review if needed.

Reply:
I prefer the second approach too, with the exception that non-static "OneTime" methods cause the fixture to error and require the user to remove the "OneTime" part or choose another resolution.